lib/elixir/lib/inspect/algebra.ex

```diff
@@ -478,24 +478,34 @@ defmodule Inspect.Algebra do
     "..."
   end
 
-  defp do_surround_many([h], limit, opts, fun, _sep) do
-    fun.(h, %{opts | limit: limit})
+  defp do_surround_many([], _limit, _opts, _fun, _sep) do
+    :doc_nil
   end
 
-  defp do_surround_many([h|t], limit, opts, fun, sep) when is_list(t) do
-    limit = decrement(limit)
-    glue(
-      concat(fun.(h, %{opts | limit: limit}), sep),
-      do_surround_many(t, limit, opts, fun, sep)
-    )
+  defp do_surround_many([h], limit, opts, fun, sep) do
+    fun.(h, %{opts | limit: limit})
   end
 
-  defp do_surround_many([h|t], limit, opts, fun, _sep) do
+  defp do_surround_many([h|t], limit, opts, fun, sep) do
     limit = decrement(limit)
-    glue(
-      concat(fun.(h, %{opts | limit: limit}), @tail_separator),
-      fun.(t, %{opts | limit: limit})
-    )
+    case fun.(h, %{opts | limit: limit}) do
```

Comment:
This seems a bit messy... maybe we can check on the `do_surround_many` clause if the h is empty and avoid calling the function body altogether?